contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -3435,3 +3510,415 @@ objectproperty_internal(PG_FUNCTION_ARGS)
 
 	PG_RETURN_NULL();
 }
+
+int 
+SPI_execute_raw_parsetree(RawStmt *parsetree, bool read_only, long tcount)
+{
+	_SPI_plan			plan;
+	int					ret;
+	List				*plancache_list;
+	CachedPlanSource	*plansource;
+	int					prev_sql_dialect;
+
+	if (parsetree == NULL || tcount < 0)
+		return SPI_ERROR_ARGUMENT;
+	
+	/*
+	 * set sql_dialect to tsql, which is needed for raw parsetree parsing 
+	 * and processing
+	 */
+	prev_sql_dialect = sql_dialect;
+	sql_dialect = SQL_DIALECT_TSQL;
+	
+	memset(&plan, 0, sizeof(_SPI_plan));
+	plan.magic = _SPI_PLAN_MAGIC;
+	plan.parse_mode = RAW_PARSE_DEFAULT;
+	plan.cursor_options = CURSOR_OPT_PARALLEL_OK;
+
+	/*
+	 * Construct plancache entries, but don't do parse analysis yet.
+	 */
+	plancache_list = NIL;
+
+	/* 
+	 * src sql can be optained from pstate->p_sourcetext, but
+	 * it is not important here
+	 */
+	plansource = CreateOneShotCachedPlan(parsetree,
+										"SQL NOT AVAILABLE",
+										CreateCommandTag(parsetree->stmt));
+
+	plancache_list = lappend(plancache_list, plansource);
+	plan.plancache_list = plancache_list;
+	plan.oneshot = true;
+
+	ret = SPI_execute_plan_with_paramlist(&plan, NULL, read_only, tcount);
+
+	/* reset sql_dialect */
+	sql_dialect = prev_sql_dialect;
+
```

Comment:
if `SPI_execute_plan_with_paramlist` hits an ERROR then you wouldnt be able to reset sql_dialect to the original value, unless you surround this with `PG_TRY` and `CATCH`. Also, do we ever get here when sql_dialect != SQL_DIALECT_TSQL?